File: backend/app/schemas/planning.py
"""API schemas for the second-stage relational planning layer."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class PlanningOperationCommand(BaseModel):
    operation_key: str = Field(min_length=8, max_length=128, pattern=r"^[A-Za-z0-9._:-]+$")


class PlanningMutationCommand(PlanningOperationCommand):
    expected_structure_version: int = Field(ge=1)


class PlanningPartCreate(PlanningMutationCommand):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=10_000)

    @field_validator("title")
    @classmethod
    def strip_title(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("篇章名称不能为空")
        return clean


class PlanningPartUpdate(PlanningMutationCommand):
    expected_lock_version: int = Field(ge=1)
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=10_000)

    @field_validator("title")
    @classmethod
    def strip_title(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("篇章名称不能为空")
        return clean


class PlanningChapterCreate(PlanningMutationCommand):
    title: str = Field(min_length=1, max_length=200)
    summary: str = Field(default="", max_length=20_000)
    target_word_count: int | None = Field(default=None, ge=500, le=10_000)

    @field_validator("title")
    @classmethod
    def strip_title(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("章节名称不能为空")
        return clean


class PlanningChapterUpdate(PlanningMutationCommand):
    expected_lock_version: int = Field(ge=1)
    title: str | None = Field(default=None, min_length=1, max_length=200)
    summary: str | None = Field(default=None, max_length=20_000)
    target_word_count: int | None = Field(default=None, ge=500, le=10_000)
    clear_target_word_count: bool = False

    @field_validator("title")
    @classmethod
    def strip_title(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not (clean := value.strip()):
            raise ValueError("章节名称不能为空")
        return clean

    @model_validator(mode="after")
    def require_update(self):
        if (
            self.title is None
            and self.summary is None
            and self.target_word_count is None
            and not self.clear_target_word_count
        ):
            raise ValueError("至少需要修改一个章节字段")
        if self.target_word_count is not None and self.clear_target_word_count:
            raise ValueError("不能同时设置并清空目标字数")
        return self


class PlanningReorderPart(BaseModel):
    part_id: str = Field(min_length=1, max_length=32)
    chapter_ids: list[str] = Field(default_factory=list, max_length=1000)


class PlanningStructureReorder(PlanningMutationCommand):
    parts: list[PlanningReorderPart] = Field(default_factory=list, max_length=1000)

    @model_validator(mode="after")
    def limit_total_nodes(self):
        if sum(len(part.chapter_ids) for part in self.parts) > 1000:
            raise ValueError("一次重排最多包含 1000 个活动章节")
        if any(len(chapter_id) > 32 for part in self.parts for chapter_id in part.chapter_ids):
            raise ValueError("章节 ID 格式无效")
        return self


class PlanningMutationReceipt(BaseModel):
    receipt_kind: Literal["structure"] = "structure"
    receipt_id: str
    operation_key: str
    operation_type: str
    replayed: bool
    changed: bool
    project_id: str
    plan_id: str
    previous_structure_version: int
    new_structure_version: int
    affected_node: dict[str, Any] | None = None
    placement: dict[str, Any] | None = None
    structure: dict[str, Any] | None = None
    created_at: datetime

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentCreate(PlanningOperationCommand):
    expected_assignment_version: int = Field(ge=1)
    element_id: str = Field(min_length=32, max_length=32)
    expected_element_content_version: int = Field(ge=1)
    scope_type: Literal["novel", "part", "chapter"]
    scope_target_id: str = Field(min_length=32, max_length=32)


class PlanningAssignmentCommand(PlanningOperationCommand):
    expected_assignment_version: int = Field(ge=1)
    expected_lock_version: int = Field(ge=1)
    scope_type: Literal["novel", "part", "chapter"]
    scope_target_id: str = Field(min_length=32, max_length=32)


class PlanningAssignmentTypeSnapshot(BaseModel):
    id: str
    key: str
    display_name: str
    status: Literal["active", "archived"]

    model_config = ConfigDict(extra="forbid")


class PlanningAssignedElementSnapshot(BaseModel):
    id: str
    name: str
    summary: str
    type: PlanningAssignmentTypeSnapshot
    confirmation_status: Literal["candidate", "confirmed", "rejected"]
    lifecycle_status: Literal["active", "archived", "merged"]
    enabled: bool
    merged_into_element_id: str | None = None

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentScopeSnapshot(BaseModel):
    scope_type: Literal["novel", "part", "chapter"]
    scope_target_id: str
    title: str
    status: Literal["active", "archived"]
    part_id: str | None = None

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentSnapshot(BaseModel):
    id: str
    element_id: str
    scope: PlanningAssignmentScopeSnapshot
    status: Literal["active", "removed"]
    lock_version: int = Field(ge=1)
    assigned_at_content_version: int = Field(ge=1)
    current_content_version: int = Field(ge=1)
    content_changed_since_assignment: bool
    element: PlanningAssignedElementSnapshot
    generation_eligible: bool
    ineligible_reasons: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentSource(BaseModel):
    assignment_id: str
    scope: PlanningAssignmentScopeSnapshot
    lock_version: int = Field(ge=1)
    assigned_at_content_version: int = Field(ge=1)

    model_config = ConfigDict(extra="forbid")


class PlanningEffectiveElementResponse(BaseModel):
    element_id: str
    current_content_version: int = Field(ge=1)
    content_changed_since_any_assignment: bool
    element: PlanningAssignedElementSnapshot
    direct_assignments: list[PlanningAssignmentSource] = Field(default_factory=list)
    inherited_from: list[PlanningAssignmentSource] = Field(default_factory=list)
    all_sources: list[PlanningAssignmentSource] = Field(default_factory=list)
    generation_eligible: bool
    ineligible_reasons: list[str] = Field(default_factory=list)

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentCounts(BaseModel):
    direct: int = Field(ge=0)
    direct_active: int = Field(ge=0)
    direct_removed: int = Field(ge=0)
    effective: int = Field(ge=0)
    generation_eligible: int = Field(ge=0)
    ineligible: int = Field(ge=0)

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentMutationReceipt(BaseModel):
    receipt_kind: Literal["assignment"] = "assignment"
    receipt_id: str
    operation_key: str
    operation_type: Literal[
        "assignment_create", "assignment_remove", "assignment_restore"
    ]
    replayed: bool
    changed: bool
    project_id: str
    plan_id: str
    previous_assignment_version: int
    new_assignment_version: int
    assignment: PlanningAssignmentSnapshot
    event_id: str
    created_at: datetime

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentScopeResponse(BaseModel):
    scope: PlanningAssignmentScopeSnapshot
    assignment_version: int
    direct_assignments: list[PlanningAssignmentSnapshot] = Field(default_factory=list)
    effective_elements: list[PlanningEffectiveElementResponse] = Field(default_factory=list)
    counts: PlanningAssignmentCounts

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentEventResponse(BaseModel):
    id: str
    action: Literal["assign", "remove", "restore"]
    previous_status: Literal["active", "removed"] | None = None
    new_status: Literal["active", "removed"]
    previous_lock_version: int = Field(ge=0)
    new_lock_version: int = Field(ge=1)
    element_content_version: int = Field(ge=1)
    performed_by: str
    created_at: datetime

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentHistoryItem(BaseModel):
    id: str
    scope: PlanningAssignmentScopeSnapshot
    status: Literal["active", "removed"]
    lock_version: int = Field(ge=1)
    events: list[PlanningAssignmentEventResponse] = Field(default_factory=list)

    model_config = ConfigDict(extra="forbid")


class PlanningAssignmentHistoryResponse(BaseModel):
    element_id: str
    assignments: list[PlanningAssignmentHistoryItem] = Field(default_factory=list)

    model_config = ConfigDict(extra="forbid")


class PlanningChapterResponse(BaseModel):
    id: str
    project_id: str
    plan_id: str
    part_id: str
    title: str
    summary: str
    target_word_count: int | None
    position: int
    status: Literal["active", "archived"]
    lock_version: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PlanningPartResponse(BaseModel):
    id: str
    project_id: str
    plan_id: str
    title: str
    description: str
    position: int
    status: Literal["active", "archived"]
    lock_version: int
    created_at: datetime
    updated_at: datetime
    chapters: list[PlanningChapterResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class NovelPlanResponse(BaseModel):
    id: str
    project_id: str
    status: Literal["active", "archived"]
    structure_version: int
    assignment_version: int
    created_at: datetime
    updated_at: datetime
    parts: list[PlanningPartResponse] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/planning.ts
export type PlanningNodeStatus = "active" | "archived";

export interface PlanningChapter {
  id: string;
  project_id: string;
  plan_id: string;
  part_id: string;
  title: string;
  summary: string;
  target_word_count: number | null;
  position: number;
  status: PlanningNodeStatus;
  lock_version: number;
  created_at: string;
  updated_at: string;
}

export interface PlanningPart {
  id: string;
  project_id: string;
  plan_id: string;
  title: string;
  description: string;
  position: number;
  status: PlanningNodeStatus;
  lock_version: number;
  created_at: string;
  updated_at: string;
  chapters: PlanningChapter[];
}

export interface NovelPlan {
  id: string;
  project_id: string;
  status: PlanningNodeStatus;
  structure_version: number;
  assignment_version: number;
  created_at: string;
  updated_at: string;
  parts: PlanningPart[];
}

export interface PlanningOperationCommand {
  operation_key: string;
  expected_structure_version: number;
}

export interface PlanningPartCreateInput extends PlanningOperationCommand {
  title: string;
  description: string;
}

export interface PlanningPartUpdateInput extends PlanningPartCreateInput {
  expected_lock_version: number;
}

export interface PlanningChapterCreateInput extends PlanningOperationCommand {
  title: string;
  summary: string;
  target_word_count: number | null;
}

export interface PlanningChapterUpdateInput extends PlanningOperationCommand {
  expected_lock_version: number;
  title?: string;
  summary?: string;
  target_word_count?: number;
  clear_target_word_count?: boolean;
}

export interface PlanningNodeStateInput extends PlanningOperationCommand {}

export interface PlanningReorderInput extends PlanningOperationCommand {
  parts: Array<{ part_id: string; chapter_ids: string[] }>;
}

export interface PlanningMutationReceipt {
  receipt_kind: "structure";
  receipt_id: string;
  operation_key: string;
  operation_type: string;
  replayed: boolean;
  changed: boolean;
  project_id: string;
  plan_id: string;
  previous_structure_version: number;
  new_structure_version: number;
  affected_node: Record<string, unknown> | null;
  placement: Record<string, unknown> | null;
  structure: Record<string, unknown> | null;
  created_at: string;
}

export type PlanningScopeType = "novel" | "part" | "chapter";

export interface PlanningScopeSnapshot {
  scope_type: PlanningScopeType;
  scope_target_id: string;
  title: string;
  status: PlanningNodeStatus;
  part_id: string | null;
}

export interface PlanningAssignmentTypeSnapshot {
  id: string;
  key: string;
  display_name: string;
  status: "active" | "archived";
}

export interface PlanningAssignedElementSnapshot {
  id: string;
  name: string;
  summary: string;
  type: PlanningAssignmentTypeSnapshot;
  confirmation_status: "candidate" | "confirmed" | "rejected";
  lifecycle_status: "active" | "archived" | "merged";
  enabled: boolean;
  merged_into_element_id: string | null;
}

export interface PlanningAssignmentSnapshot {
  id: string;
  element_id: string;
  scope: PlanningScopeSnapshot;
  status: "active" | "removed";
  lock_version: number;
  assigned_at_content_version: number;
  current_content_version: number;
  content_changed_since_assignment: boolean;
  element: PlanningAssignedElementSnapshot;
  generation_eligible: boolean;
  ineligible_reasons: string[];
  created_at: string;
  updated_at: string;
}

export interface PlanningAssignmentSource {
  assignment_id: string;
  scope: PlanningScopeSnapshot;
  lock_version: number;
  assigned_at_content_version: number;
}

export interface PlanningEffectiveElement {
  element_id: string;
  current_content_version: number;
  content_changed_since_any_assignment: boolean;
  element: PlanningAssignedElementSnapshot;
  direct_assignments: PlanningAssignmentSource[];
  inherited_from: PlanningAssignmentSource[];
  all_sources: PlanningAssignmentSource[];
  generation_eligible: boolean;
  ineligible_reasons: string[];
}

export interface PlanningAssignmentCounts {
  direct: number;
  direct_active: number;
  direct_removed: number;
  effective: number;
  generation_eligible: number;
  ineligible: number;
}

export interface PlanningAssignmentScopeResponse {
  scope: PlanningScopeSnapshot;
  assignment_version: number;
  direct_assignments: PlanningAssignmentSnapshot[];
  effective_elements: PlanningEffectiveElement[];
  counts: PlanningAssignmentCounts;
}

export interface PlanningAssignmentCreateInput {
  operation_key: string;
  expected_assignment_version: number;
  element_id: string;
  expected_element_content_version: number;
  scope_type: PlanningScopeType;
  scope_target_id: string;
}

export interface PlanningAssignmentStateInput {
  operation_key: string;
  expected_assignment_version: number;
  expected_lock_version: number;
  scope_type: PlanningScopeType;
  scope_target_id: string;
}

export interface PlanningAssignmentEvent {
  id: string;
  action: "assign" | "remove" | "restore";
  previous_status: "active" | "removed" | null;
  new_status: "active" | "removed";
  previous_lock_version: number;
  new_lock_version: number;
  element_content_version: number;
  performed_by: string;
  created_at: string;
}

export interface PlanningAssignmentHistoryItem {
  id: string;
  scope: PlanningScopeSnapshot;
  status: "active" | "removed";
  lock_version: number;
  events: PlanningAssignmentEvent[];
}

export interface PlanningAssignmentHistoryResponse {
  element_id: string;
  assignments: PlanningAssignmentHistoryItem[];
}

export interface PlanningAssignmentMutationReceipt {
  receipt_kind: "assignment";
  receipt_id: string;
  operation_key: string;
  operation_type: "assignment_create" | "assignment_remove" | "assignment_restore";
  replayed: boolean;
  changed: boolean;
  project_id: string;
  plan_id: string;
  previous_assignment_version: number;
  new_assignment_version: number;
  assignment: PlanningAssignmentSnapshot;
  event_id: string;
  created_at: string;
}

export type PlanningOperationReceipt =
  | PlanningMutationReceipt
  | PlanningAssignmentMutationReceipt;
